Add QuickCheck coverage for 2I features.

seancribbs opened the pull request:
We didn't have decent test coverage of 2I for backends. This gives a standard suite to run against the backends that currently support it, including LevelDB and Memory (and Multi if configured properly), thanks to QuickCheck.

Commit:
Add Quickcheck coverage for 2I features.

Vagabond commented on test/backend_eqc.erl:
```diff
@@ -271,17 +351,20 @@ stopped(#qcst{backend=Backend,
     ].
 
+precondition(_From,_To,#qcst{backend=Backend},{call, _M, fold_keys, [_FoldFun, _Acc, [{index, Bucket, _}], BeState]}) ->
+    {ok, Capabilities} = Backend:capabilities(Bucket, BeState),
```
This seems to call Backend:capabilities/2 with the symbolic state, and thus break the test.

Vagabond commented:
So, this test fails for me on both multi and eleveldb backends, is this expected? Memory seems to actually pass.

seancribbs commented:
@Vagabond Thanks for looking, I will debug tomorrow.

Commit:
Fix failures related to backend state.

* You can't call Backend:capabilities/2 with symbolic state, so moved the check to
  a dynamic_precondition.
* When indexes were left behind (i.e. IndexSpecs not given on delete),
  riak_kv_eleveldb_backend would report false for is_empty/1. Changed
  the direct backend call into a local call so we can determine which
  index entries to remove from the model state and then proxy the call
  into the backend.

seancribbs commented:
@Vagabond Another look-see? Thanks!

Vagabond commented:
I haz counterexample:

https://gist.github.com/5d46a79ee2af2390c6fe

This is eleveldb without async folds enabled.

Vagabond commented:
Found another with async folds on.

Vagabond commented:
You need to run quite a few tests before you'll hit one, though. The default of 50 is nowhere near what is needed.

seancribbs commented:
I will double-check the postconditions tomorrow, what it shrank to doesn't seem erroneous.

On Apr 24, 2012, at 10:26 PM, Andrew Thompsonreply@reply.github.com wrote:

> I haz counterexample:
> 
> https://gist.github.com/5d46a79ee2af2390c6fe
> 
> This is eleveldb without async folds enabled.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/basho/riak_kv/pull/315#issuecomment-5322309

seancribbs commented:
Yes, but I have run all backends with 250 or more and not provoked that example. Needs more time I guess.

Sean Cribbs

On Apr 24, 2012, at 10:41 PM, Andrew Thompsonreply@reply.github.com wrote:

> You need to run quite a few tests before you'll hit one, though. The default of 50 is nowhere near what is needed.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/basho/riak_kv/pull/315#issuecomment-5322431

Vagabond commented:
Perhaps your preconditions aren't strict enough to let it shrink correctly. Although that first countexample doesn't look like a bad shrink to me...

Commit:
Fix predicate error in remove_indexes/3.

Before this change, the model state was having all indexes
corresponding to the deleted key (in any bucket) or bucket (with any
key) removed. This meant that when index specs were being calculated
for removal in delete/5, sometimes none were there to remove,
resulting in the backend not being empty after all keys were deleted.

Commit:
Fix eunit timeout errors in multi backend.

Vagabond committed:
Ensure the EQC test cleans up backends, fix ETS leak in memory backend

Vagabond commented:
+1 to merge, now.
